google-cloud-spanner/lib/google/cloud/spanner/convert.rb

```diff
@@ -164,6 +103,64 @@ def raw_to_value obj
             end
           end
 
+          def field_for_raw obj
+            if NilClass === obj
+              raise ArgumentError, "Cannot determine type for nil values."
+            elsif String === obj || Symbol === obj
+              :STRING
+            elsif TrueClass === obj || FalseClass === obj
+              :BOOL
+            elsif Integer === obj
+              :INT64
+            elsif Numeric === obj # Any number not an integer gets to be a float
+              :FLOAT64
+            elsif Time === obj || DateTime === obj
+              :TIMESTAMP
+            elsif Date === obj
+              :DATE
+            elsif Array === obj
+              if obj.empty?
+                raise ArgumentError,
+                      "Cannot determine type for empty array values."
+              end
+              non_nil_fields = obj.compact.map { |e| field_for_raw e }.compact
+              if non_nil_fields.empty?
+                raise ArgumentError,
+                      "Cannot determine type for array of nil values."
+              end
+              if non_nil_fields.uniq.count > 1
+                raise ArgumentError,
+                      "Cannot determine type for array of different values."
+              end
+              [non_nil_fields.first]
+            elsif Hash === obj
+              raw_type_pairs = obj.map do |key, value|
+                [key, field_for_raw(value)]
+              end
+              Fields.new Hash[raw_type_pairs]
+            elsif Data === obj
+              obj.fields
+            elsif obj.respond_to?(:read) && obj.respond_to?(:rewind)
+              :BYTES
+            else
+              raise ArgumentError,
+                    "Cannot determine type for #{obj.class} values."
+            end
+          end
+
+          def type_for_field field
+            return field.to_grpc_type if field.respond_to? :to_grpc_type
+
+            if Array === field
+              Google::Spanner::V1::Type.new(
+                code: :ARRAY,
+                array_element_type: type_for_field(field.first)
+              )
+            else
+              Google::Spanner::V1::Type.new(code: field)
```

Comment:
Guard statement is preferred here than a one statement condition branching ..
something like:

```
return Google::Spanner::V1::Type.new(
                code: :ARRAY,
                array_element_type: type_for_field(field.first)
              ) if Array === field
Google::Spanner::V1::Type.new(code: field)
```

Reply:
Due to the amount of code in each conditional I'm fine keeping this method the way it is.